### PR TITLE
Added support for customization of HttpTracing

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
with this change you can define beans that will get added / set to the `HttpTracing` bean

fixes gh-886

### CODE REVIEW

File: docs/src/main/asciidoc/spring-cloud-sleuth.adoc [hunk 1]
```diff
@@ -762,11 +762,13 @@ Running the preceding method with a value of `15` leads to setting a tag with a
 
 //=== Spring Integration
 
-// TODO: Update this
-
-// === HTTP
+=== HTTP
 
-// TODO: Update this
+If a customization of client / server parsing of the HTTP related spans is required,
```
Comment (adriancole) [inline docs/src/main/asciidoc/spring-cloud-sleuth.adoc]: I might say what you have and link to brave's docs for more

ex https://github.com/openzipkin/brave/tree/master/instrumentation/http#sampling-policy shows how to make a path-based sampler

File: docs/src/main/asciidoc/spring-cloud-sleuth.adoc [hunk 2]
```diff
@@ -762,11 +762,13 @@ Running the preceding method with a value of `15` leads to setting a tag with a
...
+just register a bean of type `brave.http.HttpClientParser` or
+`brave.http.HttpServerParser`. If client /server sampling is required, just
+register a bean of type `brave.http.HttpSampler` and name the bean
+ `sleuthClientSampler` for client sampler and `sleuthServerSampler` for server sampler.
```
Comment (adriancole) [inline docs/src/main/asciidoc/spring-cloud-sleuth.adoc]: do you mean name the bean or use the annotation?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java [hunk 3]
```diff
@@ -39,28 +48,122 @@
...
+	}
+
+	@Bean
+	@ConditionalOnMissingBean(name = "sleuthClientSampler")
+	HttpSampler sleuthClientSampler() {
+		return HttpSampler.TRACE_ID;
+	}
+}
+
+class CompositeClientParser extends HttpClientParser {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java]: composite is trickier than just allowing someone to override the type. For example, they can't remove tags this way. I might start with just letting people override it. If composition becomes common we can add

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java [hunk 4]
```diff
@@ -39,28 +48,122 @@
...
+		this.parsers.forEach(parser -> parser.request(adapter, req, customizer));
+	}
+
+	@Override public <Resp> void response(HttpAdapter<?, Resp> adapter, Resp res,
+			Throwable error, SpanCustomizer customizer) {
+		this.parsers.forEach(parser -> parser.response(adapter, res, error, customizer));
+	}
 }
+
+class CompositeServerSampler extends HttpSampler {
```
Comment (adriancole) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceHttpAutoConfiguration.java]: Personally I think the skipPattern sampler is fantastically easy to override. Just have folks use the property as opposed to composing over it?

Ex SkipPattern is public so someone can easily include it in their own sampler.:
```java
class MyHttpSampler extends HttpSampler {

	private final Pattern pattern;

	SleuthHttpSampler(SkipPatternProvider provider) {
		this.pattern = provider.skipPattern();
	}

	@Override public <Req> Boolean trySample(HttpAdapter<Req, ?> adapter, Req request) {
		String url = adapter.path(request);
		boolean shouldSkip = this.pattern.matcher(url).matches();
		if (shouldSkip) {
			return false;
		}
		return null;
	}
}
```

**Comment by adriancole** [comment]
Added a commit which is mostly just polish. I do think we'll eventually get a request to not use the sleuth sampling pattern, but then again, this can also be done by overriding the pattern. So basically sounds good to me.

**Comment by marcingrzejszczak** [comment]
Fixed via https://github.com/spring-cloud/spring-cloud-sleuth/commit/375f244279265281c8fe91894522e5641ebcb7cb

**Pull request closed without merging**